programs/eosio-launcher/main.cpp

```diff
@@ -150,19 +153,19 @@ class host_def {
   vector<eosd_def> instances;
 
   uint16_t p2p_port() {
-    return base_p2p_port + p2p_count++;
+    return cluster_id * cluster_port_offset + base_p2p_port + p2p_count++;
```

Comment:
I think the way ports are calculated should be updated to reflect the limits imposed by `cluster_port_offset` and the fact that the port number must be a 16-bit unsigned integer.

I think ports could be calculated with the help of the following code:

```cpp
enum class port_type : uint8_t {
   p2p,
   http,
   NUM_PORT_TYPES
};

namespace port_constants {
   static constexpr uint16_t base = 1536;
   static constexpr uint16_t cluster_stride = 2000;
   static constexpr uint8_t  node_stride = 5;
   // maximum number of different types of ports = node_stride

   static_assert( cluster_stride > 0 );
   static_assert( base + cluster_stride <= std::numeric_limits<uint16_t>::max() );

   static constexpr uint16_t max_num_clusters = (std::numeric_limits<uint16_t>::max() - base) / cluster_stride;
   // maximum number of clusters = max_num_clusters

   static_assert( node_stride > 0 );
   static_assert( node_stride <= cluster_stride );
   static_assert( node_stride >= static_cast<uint8_t>(port_type::NUM_PORT_TYPES) );

   static constexpr uint16_t max_num_nodes_per_cluster = cluster_stride / node_stride;
   // maximum number of nodes (including bios node) per cluster = max_num_nodes_per_cluster
}

constexpr uint16_t calculate_port( uint16_t cluster_index, uint16_t node_index, port_type t ) {
   if( cluster_index >= port_constants::max_num_clusters )
     throw std::out_of_range( "invalid cluster index" );
   if( node_index >= port_constants::max_num_nodes_per_cluster )
     throw std::out_of_range( "invalid node index" );
    
   return port_constants::base 
            + (port_constants::cluster_stride * cluster_index)
            + (port_constants::node_stride * node_index)
            + static_cast<uint16_t>(t);
}
```

So for example the `http_bios_port()` would return `calculate_port( cluster_id, 0, port_type::http )` which given the example port constants defined above would be equal to 1537 if `cluster_id == 0` and 3537 if `cluster_id == 1`.

The numbers defined above for `port_constants::base`, `port_constants::cluster_stride`, and `port_constants::node_stride` are just examples. But with those example values, this system could support a maximum of 32 (`max_num_clusters`) simultaneous clusters on a single machine and a maximum of 400 (`max_num_nodes_per_cluster`) nodes per each cluster (or a maximum of 399 nodes if you don't count the bios node). And it could do this while also supporting an expansion of the number of port types from the current two (`p2p` and `http`) to 5 to accommodate other ports we may want in future tests, e.g. the state history plugin. (Also what about the keosd port? Should a separate keosd run for each cluster, or should multiple clusters running on a single machine share a single keosd instance?)